fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateGroupByKey.java

```diff
@@ -0,0 +1,217 @@
+import org.apache.doris.nereids.trees.expressions.NamedExpression;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.plans.logical.LogicalAggregate;
+import org.apache.doris.nereids.trees.plans.logical.LogicalPlan;
+import org.apache.doris.qe.ConnectContext;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Eliminate group by key based on fd item information.
+ */
+public class EliminateGroupByKey extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalAggregate(logicalProject()).then(agg -> {
+            Set<Integer> enableNereidsRules = ConnectContext.get().getSessionVariable().getEnableNereidsRules();
+            if (!enableNereidsRules.contains(RuleType.ELIMINATE_GROUP_BY_KEY.type())) {
+                return null;
+            }
+            LogicalPlan childPlan = agg.child();
+            List<FdItem> uniqueFdItems = new ArrayList<>();
+            List<FdItem> nonUniqueFdItems = new ArrayList<>();
+            if (agg.getGroupByExpressions().isEmpty()
+                    || agg.getGroupByExpressions().equals(agg.getOutputExpressions())
+                    || !agg.getGroupByExpressions().stream().allMatch(e -> e instanceof SlotReference)
+                    || agg.getGroupByExpressions().stream().allMatch(e ->
+                        ((SlotReference) e).getColumn().isPresent() && ((SlotReference) e).getTable().isPresent())) {
+                return null;
+            }
+            ImmutableSet<FdItem> fdItems = childPlan.getLogicalProperties().getFunctionalDependencies().getFdItems();
+            if (fdItems.isEmpty()) {
+                return null;
+            }
+            List<SlotReference> candiExprs = agg.getGroupByExpressions().stream()
+                    .map(SlotReference.class::cast).collect(Collectors.toList());
+
+            fdItems.stream().filter(e -> !e.isCandidate()).forEach(e -> {
+                        if (e.isUnique()) {
+                            uniqueFdItems.add(e);
+                        } else {
+                            nonUniqueFdItems.add(e);
+                        }
+                    }
+            );
+
+            int minParentExprCnt = -1;
+            ImmutableSet<SlotReference> minParentExprs = ImmutableSet.of();
+            // if unique fd items exists, try to find the one which has the
+            // smallest parent exprs
+            for (int i = 0; i < uniqueFdItems.size(); i++) {
+                FdItem fdItem = uniqueFdItems.get(i);
+                ImmutableSet<SlotReference> parentExprs = fdItem.getParentExprs();
+                if (minParentExprCnt == -1 || parentExprs.size() < minParentExprCnt) {
+                    boolean isContain = isExprsContainFdParent(candiExprs, fdItem);
```

Comment:
Is it redundant? If there are unique slots, we can always use one unique slot to replace all group by keys